File: src/main/java/org/radarcns/android/auth/portal/SubjectTokenParser.java
package org.radarcns.android.auth.portal;

import android.support.annotation.NonNull;
import org.radarcns.android.auth.AppAuthState;
import org.radarcns.android.auth.AuthStringParser;

import java.io.IOException;

public class SubjectTokenParser implements AuthStringParser {
    private final AccessTokenParser accessTokenParser;
    private final ManagementPortalClient client;

    public SubjectTokenParser(ManagementPortalClient client, AppAuthState state) {
        this.client = client;
        this.accessTokenParser = new AccessTokenParser(state);
    }

    @Override
    public AppAuthState parse(@NonNull String body) throws IOException {
        AppAuthState newState = this.accessTokenParser.parse(body);
        return client.getSubject(newState, new GetSubjectParser(newState));
    }
}

File: radar-commons-android/src/main/java/org/radarcns/android/auth/portal/SubjectTokenParser.kt
package org.radarcns.android.auth.portal

import org.radarcns.android.auth.AppAuthState
import org.radarcns.android.auth.AuthStringParser

import java.io.IOException

class SubjectTokenParser(private val client: ManagementPortalClient, state: AppAuthState) : AuthStringParser {
    private val accessTokenParser: AccessTokenParser = AccessTokenParser(state)

    @Throws(IOException::class)
    override fun parse(body: String): AppAuthState {
        val newState = this.accessTokenParser.parse(body)
        return client.getSubject(newState, GetSubjectParser(newState))
    }
}
